### ISSUE TITLE
Update MongoDB driver to v6

### ISSUE THREAD

**Issue by StorytellerCZ** [meteor/meteor#13218]
So just noticed that we are two major versions behind the MongoDB Node driver.
Currently on 4.17.2. I would like to bring this up for an upgrade.

I think the reason why this has not been done is that [v5](https://github.com/mongodb/node-mongodb-native/releases/tag/v5.0.0) of the package has removed callbacks, which is an issue for Meteor 2.x and bellow, but I think that for Meteor 3 it should be fine.

So steps for upgrade going forward:
## [5.0.1](https://github.com/mongodb/node-mongodb-native/releases/tag/v5.0.0)
* Remove any remaining callbacks from server or make a workaround with the compatibility package

## [5.1.0](https://github.com/mongodb/node-mongodb-native/releases/tag/v5.1.0)
* Check if we need to do anything with the new `bigint`

## [5.2.0](https://github.com/mongodb/node-mongodb-native/releases/tag/v5.2.0)
Doesn't seem like we need to do anything here.

## [5.3.0](https://github.com/mongodb/node-mongodb-native/releases/tag/v5.3.0)
* cursor forEach has been deprecated, we need to do the same, I think this is going to be huge as I have seen quiet a few things build on top of this.

## [5.4.0](https://github.com/mongodb/node-mongodb-native/releases/tag/v5.4.0)
More deprecations here, but nothing that should impact us,  I think.

## [5.5.0](https://github.com/mongodb/node-mongodb-native/releases/tag/v5.5.0)
Nothing of note

## [5.6.0](https://github.com/mongodb/node-mongodb-native/releases/tag/v5.6.0)
* Node 20 support
I think this is the initial release that we should really target.

## [5.7.0](https://github.com/mongodb/node-mongodb-native/releases/tag/v5.7.0)
* Lot of deprecations
* Support for change stream split events - I think stuff like this will be become crucial for moving of oplog tailing
* Search indexes

## [5.8.0](https://github.com/mongodb/node-mongodb-native/releases/tag/v5.8.0)
Should be fine

## [5.9.0](https://github.com/mongodb/node-mongodb-native/releases/tag/v5.9.0)
* Fixes memory leak with change streams

## [6.0.0](https://github.com/mongodb/node-mongodb-native/releases/tag/v6.0.0)
* Minimum Node v16
* We might need to install `socks` dependency
* Callbacks removed from ClientEncryption's encrypt, decrypt, and createDataKey methods
* This needs to be mentioned especially in changelog: Boolean options only accept 'true' or 'false' in connection strings and Repeated options are no longer allowed in connection strings

## [6.1.0](https://github.com/mongodb/node-mongodb-native/releases/tag/v6.1.0)
Looks fine

## [6.2.0](https://github.com/mongodb/node-mongodb-native/releases/tag/v6.2.0)
* insertedIds in bulk write now contain only successful insertions - should not affect us
* Fix for edge case `findOne` memory leak (same as 5.9.1)

## [6.3.0](https://github.com/mongodb/node-mongodb-native/releases/tag/v6.3.0)
* New client option `serverMonitoringMode` - just highlighted one of the new options that should be mentioned in changelog
* GridFS fields deprecated

## [6.4.0](https://github.com/mongodb/node-mongodb-native/releases/tag/v6.4.0)
* `countDocuments` now types the filter using the collection Schema - should have no effect
* Errors on cursor transform streams are now properly propagated - I don't expect a direct impact here, but this could affect some apps

## [6.5.0](https://github.com/mongodb/node-mongodb-native/releases/tag/v6.5.0)
* Bulk Write Operations Generate Ids using `pkFactory` - I don't think this should affect us, but good to keep in mind
* Fixed memory leak in Connection layer

## [6.6.0](https://github.com/mongodb/node-mongodb-native/releases/tag/v6.6.0)
* Aggregation pipelines can now add stages manually - no direct affect, but a useful feature
* `indexExists()` no longer supports the `full` option - hopefully no direct impact as well

## [6.7.0](https://github.com/mongodb/node-mongodb-native/releases/tag/v6.7.0)
Seems fine

## [6.8.0](https://github.com/mongodb/node-mongodb-native/releases/tag/v6.8.0)
* Cursor responses are now parsed lazily 🦥 <= potential speed improvement?

Given Meteor 3, I think v5.6.0 should be our target for initial upgrade to get started (though we can go to latest v5 if me make it that far, I think). Don't really know much, but I will try to get something out soon and see how far I can get with a simple version update.

**Comment by StorytellerCZ** [comment]
Initial attempt yields this error:
```
W20240705-18:54:53.680(2)? (STDERR) meteor://💻app/packages/mongo.js:153
W20240705-18:54:53.680(2)? (STDERR)       if (document instanceof MongoDB.ObjectID) {
W20240705-18:54:53.680(2)? (STDERR)                    ^
W20240705-18:54:53.680(2)? (STDERR) 
W20240705-18:54:53.681(2)? (STDERR) TypeError: Right-hand side of 'instanceof' is not an object
W20240705-18:54:53.681(2)? (STDERR)     at replaceMongoAtomWithMeteor (packages/mongo/mongo_driver.js:82:15)
W20240705-18:54:53.681(2)? (STDERR)     at replaceTypes (packages/mongo/mongo_driver.js:136:30)
W20240705-18:54:53.681(2)? (STDERR)     at AsynchronousCursor._nextObjectPromise (packages/mongo/mongo_driver.js:1092:13)
W20240705-18:54:53.681(2)? (STDERR)     at processTicksAndRejections (node:internal/process/task_queues:95:5)
W20240705-18:54:53.681(2)? (STDERR)     at AsynchronousCursor.forEach (packages/mongo/mongo_driver.js:1141:19)
W20240705-18:54:53.682(2)? (STDERR)     at Cursor.cursor.<computed> [as forEachAsync] (packages/matb33:collection-hooks/find.js:60:24)
W20240705-18:54:53.682(2)? (STDERR) 
W20240705-18:54:53.682(2)? (STDERR) Node.js v20.12.2
```

Trying a higher version results in other error:
```
I20240705-18:52:19.221(2)? Exception in callback of async function: TypeError: observeHandle.stop is not a function
I20240705-18:52:19.221(2)?     at boundEnvironment (packages/socialize:server-presence/server-presence.js:119:23)
I20240705-18:52:19.221(2)?     at processTicksAndRejections (node:internal/process/task_queues:95:5)
I20240705-18:52:19.221(2)?     at packages/meteor.js:1399:19
```
Though in other package it is a common pattern to use `handle.stop()` and similar functions. So we will have to be very careful about these things.

**Comment by rj-david** [comment]
Looking forward for this update @StorytellerCZ 👍🏼

But if possible, can we ensure that we have a patch upgrade for fixes only without any need for another set of breaking changes?

Since Meteor 3.0 is a very big change, it's expected that there will be bugs that's needed to be ironed out for the next version. So hopefully, there is a 3.0.1 that is targeted only for fixes. A version that we can upgrade to without the fear of breaking something again.

If the next version is 3.1, and we are expecting breaking changes again, that will be giving another set of pressure as a required upgrade (due to expected fixes).

Thanks

**Comment by StorytellerCZ** [comment]
I have intentionally put this as a target for Meteor 3.1 as it is not so urgent from the functionality perspective.
I don't think there will be any breaking changes in the end, but some internal code will have to be adjusted. If that is not the case there is a mongo package that allows for the current approach to work.

**Comment by nachocodoner** [comment]
FYI, I've started [the PR to upgrade the Mongo driver](https://github.com/meteor/meteor/pull/13429), beginning with 5.x due to major breaking changes like the removal of callbacks. Once this is done, and depending on the time left for 3.1 and migration difficulty for 6.x, we'll move forward with that.

**Comment by StorytellerCZ** [comment]
@nachocodoner it might be a good idea to upgrade the integrated MongoDB as well.

**Comment by nachocodoner** [comment]
> it might be a good idea to upgrade the integrated MongoDB as well.

That will come later, since it requires further checking.

**Comment by nachocodoner** [comment]
Mongo driver 6 has been delivered and [is available in Meteor 3.1](https://dev.to/meteor/meteorjs-31-a-new-dawn-for-full-stack-javascript-development-4a13). Pay close attention to replacing any direct usage of the driver (rawCollection/rawDatabase) and updating the code to use Promises instead of callbacks.

Update your app and share any feedback.

```shell
meteor update --release 3.1
```
